packages/wallet-adapter-react/src/WalletProvider.tsx

```diff
@@ -392,6 +407,57 @@ export const AptosWalletAdapterProvider: FC<AptosWalletProviderProps> = ({
       );
     };
   }, [wallets, account]);
+
+  // Define specific return types based on wallet type
+  type SolanaWalletDetails = { address: string; publicKey: SolanaPublicKey };
+  type EVMWalletDetails = { address: string; publicKey?: undefined };
+  type AptosWalletDetails = AccountInfo | null;
+
+  // Function overloads
+  function getOriginWalletDetails(
+    wallet: SolanaDerivedWallet
+  ): Promise<SolanaWalletDetails>;
+  function getOriginWalletDetails(
+    wallet: EIP1193DerivedWallet
+  ): Promise<EVMWalletDetails>;
+  function getOriginWalletDetails(
+    wallet: AdapterWallet
+  ): Promise<AptosWalletDetails>;
+
+  // Implementation
+  async function getOriginWalletDetails(
+    wallet: AdapterWallet
+  ): Promise<OriginWalletDetails> {
+    if (isSolanaDerivedWallet(wallet)) {
```

Comment:
why not using `instanceof` directly here?

Reply:
we can, but we already have the helper function that does exactly that